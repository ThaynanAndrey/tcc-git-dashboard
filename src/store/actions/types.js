export const GET_PULL_REQUESTS_REPOSITORY = 'GET_PULL_REQUESTS_REPOSITORY';

export const LOAD_PROJECT_PULL_REQUESTS_SUCCESS = 'LOAD_PROJECT_PULL_REQUESTS_SUCCESS';

export const LOAD_PULL_REQUESTS_NO_PROJECT_SUCCESS = 'LOAD_PULL_REQUESTS_NO_PROJECT_SUCCESS';

export const ADDED_PULL_REQUEST_SUCCESS = "ADDED_PULL_REQUEST_SUCCESS";

export const ADDED_PULL_REQUEST_ERROR = "ADDED_PULL_REQUEST_ERROR";