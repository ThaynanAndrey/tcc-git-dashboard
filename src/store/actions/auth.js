import firebase from 'firebase/app';

import { USER_LOGIN_SUCCESS, USER_LOGIN_ERROR } from './types';

export const login = () => {
    return dispatch => {
        const provider = new firebase.auth.GithubAuthProvider();
        firebase.auth().signInWithPopup(provider).then(function(result) {
            const { currentUser } = firebase.auth();
            const accessToken = result.credential.accessToken;
            dispatch({
                type: USER_LOGIN_SUCCESS,
                accessToken
            });
        }).catch(function(error) {
            console.log("erro: ", error);
            dispatch({
                type: USER_LOGIN_ERROR
            });
        });
    }
}