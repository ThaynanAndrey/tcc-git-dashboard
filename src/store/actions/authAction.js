import firebase from 'firebase/app';

import { USER_LOGIN_SUCCESS, USER_LOGIN_ERROR, USER_LOGOUT_SUCCESS, USER_LOGOUT_ERROR } from './types';

/**
 * Realize login with firebase authentication.
 * 
 * @returns {Function} Dispatch used to invoke the auth's redux
 */
export const login = () => {
    return dispatch => {
        return firebase.auth().setPersistence(firebase.auth.Auth.Persistence.SESSION)
            .then(() => {
                const provider = new firebase.auth.GithubAuthProvider();
                provider.addScope('repo');
                
                return firebase.auth().signInWithPopup(provider)
                    .then(result => {
                        const { currentUser } = firebase.auth();
                        const accessToken = result.credential.accessToken;
                        sessionStorage.setItem('authAccessToken', accessToken);
                        
                        dispatch({
                            type: USER_LOGIN_SUCCESS,
                            user: currentUser,
                            idUser: currentUser.uid,
                            accessToken
                        });
                    }).catch(error => {
                        console.log("erro: ", error);
                        dispatch({
                            type: USER_LOGIN_ERROR,
                            errorLogin: error
                        });
                    });
            }).catch(error => {
                console.log("erro: ", error);
                dispatch({
                    type: USER_LOGIN_ERROR,
                    errorLogin: error
                });
            })
    };
}

/**
 * Realize log out user.
 * 
 * @returns {Function} Dispatch used to invoke the auth's redux
 */
export const logout = () => {
    return dispatch => {
        return firebase.auth().signOut().then(() => {
            sessionStorage.removeItem('authAccessToken');
            dispatch({
                type: USER_LOGOUT_SUCCESS
            });
          }, error => {
            dispatch({
                type: USER_LOGOUT_ERROR,
                errorLogout: error
            });
          });
    };
};

/**
 * Check if exists an user authenticated on system.
 * 
 * @returns {Boolean} {@code true} if exists an user authenticated, otherwise {@code false}
 */
export const isAuthenticated = () => {
    return sessionStorage.getItem('authAccessToken') !== null;
}