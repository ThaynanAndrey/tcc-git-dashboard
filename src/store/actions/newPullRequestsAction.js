import axios from 'axios';
import firebase from 'firebase/app';

import { getRepositoriesFirestoreByIdProject } from './repositoriesAction';
import { binarySearch, mapearAtributosPullRequest, getUrlAuthenticated, getDataElemsFirestore } from '../../utils/utils';
import { LOAD_PULL_REQUESTS_NO_PROJECT_SUCCESS, ADDED_PULL_REQUEST_SUCCESS, ADDED_PULL_REQUEST_ERROR,
        LOAD_EXTERNAL_PULL_REQUESTS_NO_PROJECT_SUCCESS, LOAD_EXTERNAL_PULL_REQUESTS_NO_PROJECT_ERROR,
        RESET_PULL_REQUESTS_NO_PROJECT } from './types';

const COLLECTION_PULL_REQUESTS_FIRESTORE = "pullRequests";
const COLLECTION_PROJECTS_FIRESTORE = "projects";

const INDEX_REPOSITORIES_FIRESTORE = 0;
const INDEX_PR_FIRESTORE = 1;

/**
 * Gets all Pull Request that not added to user's Project.
 * 
 * @param {String} idProject Project id
 * @returns {Function} Dispatch used to invoke the newPullRequest's redux
 */
export const getPullRequestsNoProject = idProject => {
    return async dispatch => {
        const promises = [];
        promises.push(getRepositoriesFirestoreByIdProject(idProject));
        promises.push(_getPullRequestsFirestoreByIdProject(idProject));

        const results = await Promise.all(promises);
        const repositoriesFirestore = results[INDEX_REPOSITORIES_FIRESTORE];
        const pullRequestsFirestore = results[INDEX_PR_FIRESTORE];

        const gitHubPromises = repositoriesFirestore.map(repository => _getPullRequestsRepository(repository.idGitHub));
        const pullRequestsGitHubResult = await Promise.all(gitHubPromises);
        let nextPullRequests = await _getNextPullRequests(pullRequestsGitHubResult);
        
        let pullRequestsGit = pullRequestsGitHubResult.reduce((array, result) => array.concat(result.data), []);
        nextPullRequests = nextPullRequests.reduce((array, pr) => array.concat(pr.data), []);
        pullRequestsGit = pullRequestsGit.concat(nextPullRequests);
        pullRequestsGit = _addIdFirestoreRepository(pullRequestsGit, repositoriesFirestore);

        const pullRequestsNoProject = _getPullRequestsNoProject(pullRequestsGit, pullRequestsFirestore);

        dispatch({
            type: LOAD_PULL_REQUESTS_NO_PROJECT_SUCCESS,
            pullRequestsNoProject
        });
    };
};

/**
 * Add a new Pull Request to user's Project.
 * 
 * @param {Object} pullRequest
 *      Pull Request to be added
 * @param {String} idProject 
 *      Project id
 * @returns {Function} Dispatch used to invoke the newPullRequest's redux
 */
export const addPullRequestInProject = (pullRequest, idProject) => {
    return (dispatch, getState) => {
        const firestore = firebase.firestore();
        return firestore.collection(COLLECTION_PULL_REQUESTS_FIRESTORE).add({
            idPullRequestGitHub: pullRequest.id,
            idRepositorioGitHub: pullRequest.repositorio.id,
            nomeRepositorioGitHub: pullRequest.repositorio.nome,
            numeroPullRequestGitHub: pullRequest.numero,
            nomePropietario: pullRequest.propietario.nome,
            project: firestore.doc(`${COLLECTION_PROJECTS_FIRESTORE}/${idProject}`)
        }).then(() => {
            let pullRequestsNoProject = getState().newPullRequests.pullRequestsNoProject;
            pullRequestsNoProject = pullRequestsNoProject
                .filter(pr => pr.id !== pullRequest.id);
            
            let externalPullRequestsNoProject = getState().newPullRequests.externalPullRequestsNoProject;
            externalPullRequestsNoProject = externalPullRequestsNoProject
                .filter(pr => pr.id !== pullRequest.id);

            dispatch({
                type: ADDED_PULL_REQUEST_SUCCESS,
                pullRequestsNoProject,
                externalPullRequestsNoProject
            });
        }).catch(erro => {
            dispatch({ type: ADDED_PULL_REQUEST_ERROR , erro });
        });
    };
};

/**
 * Searches for Pull Request that aren't in the repositories added in the Project.
 * 
 * @param {String} ownerName
 *      Repository's qwner name
 * @param {String} repositoryName
 *      Repository name where are the Pull Requests searched
 * @param {String} prNumber
 *      Number Pull Request to be searched
 * @param {String} idProject
 *      Project's id
 * @returns {Function} Dispatch used to invoke the newPullRequest's redux
 */
export const searchExternalPullRequest = (ownerName, repositoryName, prNumber, idProject) => {
    return dispatch => {
        const promises = [];
        promises.push(_getGitHubPullRequests(ownerName, repositoryName, prNumber));
        promises.push(_getPullRequestsFirestoreByIdProject(idProject));

        return Promise.all(promises)
            .then(result => {
                let pullRequests = result[0];
                const isArray = pullRequests.data instanceof Array;
                let externalPullRequestsNoProject = isArray ? pullRequests.data.map(pr => mapearAtributosPullRequest(pr))
                    : new Array(mapearAtributosPullRequest(pullRequests.data));
                
                _getNextPullRequestsSimpleObject(pullRequests).then(nextPullRequests => {
                    nextPullRequests = nextPullRequests.reduce((array, pr) => array.concat(pr.data), []);
                    nextPullRequests = nextPullRequests.map(pr => mapearAtributosPullRequest(pr));
                    externalPullRequestsNoProject = externalPullRequestsNoProject.concat(nextPullRequests);

                    const projectPullRequests = result[1];

                    externalPullRequestsNoProject = externalPullRequestsNoProject.filter(externalPR =>
                        !_containsPullRequest(projectPullRequests, externalPR));
                        
                    dispatch({
                        type: LOAD_EXTERNAL_PULL_REQUESTS_NO_PROJECT_SUCCESS,
                        externalPullRequestsNoProject
                    });
                });
            })
            .catch(error => {
                dispatch({ type: LOAD_EXTERNAL_PULL_REQUESTS_NO_PROJECT_ERROR })
            });
    };
};

/**
 * Resets pull requests no Project list.
 * 
 * @returns {Function} Dispatch used to invoke the newPullRequest's redux
 */
export const resetPullRequestsNoProject = () =>
    dispatch => dispatch({ type: RESET_PULL_REQUESTS_NO_PROJECT });

/**
 * Returns a boolean indicating if the Pull Request is in pullRequests list.
 * 
 * @param {Array} pullRequests
 *      List of Pull Requests
 * @param {Object} pullRequest
 *      Pull Request to be searched
 * @return {Boolean} {@code true} if pull requests' list contains pull request, otherwise {@code false}
 */
const _containsPullRequest = (pullRequests, pullRequest) =>
    pullRequests.reduce((contains, currentPR) => 
        contains || currentPR.idPullRequestGitHub === pullRequest.id, false)

/**
 * Requests to GitHub the data's Pull Requests for: owner name, pull request name and pull request number.
 * 
 * @param {String} ownerName
 *      Owner's name
 * @param {Object} repositoryName
 *      Repository's name
 * @param {Object} prNumber
 *      Pull request number
 * @returns {Promise} request's promise
 */
const _getGitHubPullRequests = (ownerName, repositoryName, prNumber) => {
    const method = 'GET';
    const url = `https://api.github.com/repos/${ownerName}/${repositoryName}/pulls` 
        + (prNumber ? ("/" + prNumber) : "");
    const accessToken = sessionStorage.getItem('authAccessToken');
    const authUrl = getUrlAuthenticated(url, method, accessToken);

    return axios(authUrl);
};

/**
 * Requests to GitHub Repository's Pull Requests.
 * 
 * @param {Number} id
 *      ID of GitHub repository
 * @param {Number} page
 *      Pull Request's page
 * @returns {Promise} request's promise
 */
const _getPullRequestsRepository = (id, page) => {
    const method = 'GET';
    const url = `https://api.github.com/repositories/${id}/pulls?page=${page ? page : 1}`;
    const accessToken = sessionStorage.getItem('authAccessToken');
    const authUrl = getUrlAuthenticated(url, method, accessToken);

    return axios(authUrl);
};

/**
 * Requests to Firestore all Project's Pull Requests.
 * 
 * @returns {Promise} request's promise
 */
const _getPullRequestsFirestore = () => {
    const firestore = firebase.firestore();
    return firestore.collection(COLLECTION_PULL_REQUESTS_FIRESTORE).orderBy('idPullRequestGitHub').get();
};

/**
 * Filter Pull Requests by Project id.
 * 
 * @param {Array} pullRequests
 *      Pull Requests to be filtered
 * @param {String} idProject
 *      Project id
 * @returns {Array} filtered Pull Requests by Project id
 */
const _filterPullRequestsByProject = (pullRequests, idProject) =>
    pullRequests.filter(pullRequest => pullRequest.project.id === idProject);

/**
 * Gets Pull Requets' Repositories in Firestore
 * 
 * @param {String} idProject Project id
 * @returns {Array} project Pull Requests
 */
const _getPullRequestsFirestoreByIdProject = async idProject => {
    let pullRequestsFirestore = await _getPullRequestsFirestore();
    pullRequestsFirestore = getDataElemsFirestore(pullRequestsFirestore);

    return _filterPullRequestsByProject(pullRequestsFirestore, idProject);
};

/**
 * Get next pages of Pull Request.
 * 
 * @param {Array} pullRequestsGitHub
 *      Array with GitHub's Pull Requests
 * @returns {Promise} Promise with next pages os Pull Requests
 */
const _getNextPullRequests = (pullRequestsGitHub) => {
    const promises = [];
    for(const pullRequests of pullRequestsGitHub) {
        const nextPages = pullRequests.headers.link;
        if(nextPages) {
            const endPage = Number(nextPages.substring(
                nextPages.lastIndexOf("page=") + 5, 
                nextPages.lastIndexOf(">")
            ));
            for(let page = 2; page <= endPage; page++) {
                const id = pullRequests.data[0].base.repo.id;
                promises.push(_getPullRequestsRepository(id, page));
            }
        }
    }

    return Promise.all(promises);
};

/**
 * Get next pages of Pull Request.
 * 
 * @param {Array} pullRequestsGitHub
 *      Array with GitHub's Pull Requests
 * @returns {Promise} Promise with next pages os Pull Requests
 */
const _getNextPullRequestsSimpleObject = pullRequestsGitHub => {
    const promises = [];
    const nextPages = pullRequestsGitHub.headers.link;
    if(nextPages) {
        const endPage = Number(nextPages.substring(
            nextPages.lastIndexOf("page=") + 5, 
            nextPages.lastIndexOf(">")
        ));
        for(let page = 2; page <= endPage; page++) {
            const id = pullRequestsGitHub.data[0].base.repo.id;
            promises.push(_getPullRequestsRepository(id, page));
        }
    }

    return Promise.all(promises);
};

/**
 * Gets all Pull Requests with datas of Github and that isn't linked Project.
 * 
 * @param {Array} pullRequestsGit
 *      Array with Pull Requests from GitHub
 * @param {Array} pullRequestsFirestore 
 *      Array with PullRequests from Firestore
 * @returns {Array} Pull Requests' array without Project
 */
const _getPullRequestsNoProject = (pullRequestsGit, pullRequestsFirestore) => {
    return pullRequestsGit
            .filter((pullRequest) => !_isPullRequestInFirestore(pullRequest, pullRequestsFirestore))
            .map((pullRequest) => mapearAtributosPullRequest(pullRequest));
};

/**
 * Returns {@code true} if pull request is into pullRequestsFirestore, {@code false} otherwise.
 * 
 * @param {Array} pullRequest
 *      Pull Request to be verified
 * @param {Array} pullRequestsFirestore
 *      Array with Firestore's pull requests
 * @returns {@code true} if pull request is into pullRequestsFirestore, {@code false} otherwise
 */
const _isPullRequestInFirestore = (pullRequest, pullRequestsFirestore) => {
    const index = binarySearch(pullRequestsFirestore, 0, pullRequestsFirestore.length-1, pullRequest.id);
    return index !== -1;
};

/**
 * Adds Firestore id in repositories array.
 * 
 * @param {Array} pullRequestsGit 
 *      
 * @param {Array} repositoriesFirestore 
 */
const _addIdFirestoreRepository = (pullRequestsGit, repositoriesFirestore) => {
    return pullRequestsGit.map(pr => {
        const mappedPr = pr;
        mappedPr.idFirestoreRepository = repositoriesFirestore
            .filter(repository => repository.idGitHub === pr.base.repo.id)[0].id;
        
        return mappedPr;
    });
};