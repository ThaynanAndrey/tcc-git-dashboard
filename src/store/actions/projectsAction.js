import firebase from 'firebase/app';

import { getDataElemsFirestore, getFormattedDate } from '../../utils/utils';
import { LOAD_PROJECTS_SUCCESS, LOAD_PROJECTS_ERROR, CREATE_PROJECT_SUCCESS,
        CREATE_PROJECT_ERROR, GET_PROJECT_SUCCESS } from './types';

const COLLECTION_PROJECTS_FIRESTORE = "projects";
const COLLECTION_USER_FIRESTORE = "users";

/**
 * Loads user's Projects.
 * 
 * @returns {Function} dispatch used to invoke the projects' redux
 */
export const loadProjects = () => {
    return dispatch => {
        return _getPullRequestsFirestore()
            .then(projects => {
                projects = getDataElemsFirestore(projects);
                projects = _filterUserProjects(projects);
                projects = projects.map(project => {
                    let pro = project;
                    const date = new Date(pro.creationDate.seconds * 1000);
                    pro.creationDate = getFormattedDate(date);
                    return pro;
                });

                dispatch({
                    type: LOAD_PROJECTS_SUCCESS,
                    projects
                });
            })
            .catch(error => {
                dispatch({
                    type: LOAD_PROJECTS_ERROR,
                    error
                });
            });
    }
};

/**
 * Gets a project by id.
 * 
 * @param {String} idProject
 *      Project id to be gotten
 */
export const getProject = idProject =>
    dispatch => {
        const firestore = firebase.firestore();
        firestore.collection(COLLECTION_PROJECTS_FIRESTORE).doc(idProject).get()
            .then(project => {
                project = project.data()
                dispatch({ type: GET_PROJECT_SUCCESS, project })
            })
            .catch(error => console.log(error));
    }

/**
 * Creates a new Project in Firestore.
 * 
 * @param {String} projectName
 *      Project name
 * @returns {Function} dispatch used to invoke the projects' redux
 */
export const createProject = projectName => {
    return (dispatch, getState) => {
        const firestore = firebase.firestore();
        const idUser = _getIdUser();
        const creationDate = new Date();
        const newProject = {
            name: projectName,
            creationDate,
            user: firestore.doc(`${COLLECTION_USER_FIRESTORE}/${idUser}`)
        };

        return firestore.collection(COLLECTION_PROJECTS_FIRESTORE).add(newProject)
            .then(project => {
                newProject.id = project.id;
                newProject.creationDate = getFormattedDate(newProject.creationDate);
                let projects = getState().projects.projects;
                projects.push(newProject);

                dispatch({ 
                    type: CREATE_PROJECT_SUCCESS,
                    projects
                });
            }).catch(error => {
                dispatch({ type: CREATE_PROJECT_ERROR , error });
            });
    };
};

/**
 * Deletes project in Firestore.
 * 
 * @param {Object} project 
 *      Project to be deleted
 * @returns {Function} dispatch used to invoke the projects' redux
 */
export const deleteProject = (project) => {
    return (dispatch, getState) => {
        const firestore = firebase.firestore();
        return firestore.collection(COLLECTION_PROJECTS_FIRESTORE).doc(project.id).delete()
            .then(() => {
                const projects = getState().projects.projects
                    .filter(proj => proj.id !== project.id);

                dispatch({
                    type: LOAD_PROJECTS_SUCCESS,
                    projects
                });
            });
    };
};

/**
 * Gets in Firestore all user's Projects.
 * 
 * @returns {Promise} request's promise
 */
const _getPullRequestsFirestore = () => {
    const firestore = firebase.firestore();

    return firestore.collection(COLLECTION_PROJECTS_FIRESTORE).get();
};

/**
 * Filter Projects by current User.
 * 
 * @param {Array} projects
 *      Projects to be filtered
 */
const _filterUserProjects = projects => {
    const idUser = _getIdUser();
    
    return projects.filter(project => project.user.id === idUser);
};

/**
 * Gets the Firestore User id.
 * 
 * @returns {String} User id
 */
const _getIdUser = () => {
    const { currentUser } = firebase.auth();
    return currentUser.uid;
}