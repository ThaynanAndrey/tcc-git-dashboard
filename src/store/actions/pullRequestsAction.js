import axios from 'axios';
import firebase from 'firebase/app';

import { mapearAtributosPullRequest, getDataPullRequestsFirestore } from '../../utils/utils';
import { LOAD_PROJECT_PULL_REQUESTS_SUCCESS } from './types';

const PULL_REQUEST_FIRESTORE_COLLECTION = "pullRequests";

/**
 * Gets Project's Pull Requests.
 * 
 * @returns {Function} Dispatch used to invoke the pullRequest's redux
 */
export const getProjectPullRequests = () => {
    return async (dispatch, getState) => {
        let pullRequestsFirestore = await _getPullRequestsFirestore();
        pullRequestsFirestore = getDataPullRequestsFirestore(pullRequestsFirestore);
        
        const promises = pullRequestsFirestore
            .map(pullRequest => _getGitHubPullRequest(pullRequest));
        
        const pullRequestsResult = await Promise.all(promises);
        const projectPullRequests = pullRequestsResult
            .map((result, index) => 
                mapearAtributosPullRequest(result.data, pullRequestsFirestore, index));

        dispatch({
            type: LOAD_PROJECT_PULL_REQUESTS_SUCCESS,
            projectPullRequests
        });
    };
};

/**
 * Removes the Pull Request of Project.
 * 
 * @param {Object} removedPullRequest
 *      Pull Request to be removed
 * @returns {Function} Dispatch used to invoke the pullRequest's redux
 */
export const removePullRequestProject = (removedPullRequest) => {
    return (dispatch, getState) => {
        const firestore = firebase.firestore();
        return firestore.collection(PULL_REQUEST_FIRESTORE_COLLECTION).doc(removedPullRequest.idFirestore)
            .delete().then(() => {
                const projectPullRequests = getState().pullRequests.projectPullRequests
                    .filter(pullRequest => pullRequest.idFirestore !== removedPullRequest.idFirestore);

                dispatch({
                    type: LOAD_PROJECT_PULL_REQUESTS_SUCCESS,
                    projectPullRequests
                });
            });
    }
};

/**
 * Get all user's Repositories.
 * 
 * @returns {Function} Dispatch used to invoke the repositories's redux
 */
export const getAllUserRepositories = () => {
    return (dispatch, getState) => {
        const authOptions = {
            method: 'GET',
            url: 'https://api.github.com/user/repos',
            headers: {
                'Authorization': `Bearer ${getState().pullRequests.accessToken}`,
                'Content-Type': 'application/json'
            }
        };
        
        return axios(authOptions).then((response) => {
            console.log(response);
        }).catch((error) => {
            console.log(error);
        });
    }
};

/**
 * Requests to GitHub the Pull Request.
 * 
 * @param {Object} pullRequest
 *      Pull Request to be requested
 * @returns {Promise} request's promise
 */
const _getGitHubPullRequest = pullRequest => 
    axios.get(`https://api.github.com/repos/${pullRequest.nomePropietario}/${pullRequest.nomeRepositorioGitHub}/pulls/${pullRequest.numeroPullRequestGitHub}`);

/**
 * Requests to Firestore all Project's Pull Requests.
 * 
 * @returns {Promise} request's promise
 */
const _getPullRequestsFirestore = () => {
    const firestore = firebase.firestore();
    return firestore.collection(PULL_REQUEST_FIRESTORE_COLLECTION).get();
};