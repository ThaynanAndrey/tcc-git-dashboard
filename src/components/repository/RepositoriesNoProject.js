import React, { Component } from 'react';
import { connect } from 'react-redux';
import Tooltip from "react-simple-tooltip";
import { Link } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import { requireAuthentication } from '../../high-order-components/RequireAuthentication';

import { getRepositoriesNoProject, addRepositoryInProject, searchExternalRepository, resetExternalRepositories } from '../../store/actions/repositoriesAction';

const styles = {
    fontSize: "17px",
    marginRight: "50px",
    marginLeft: "50px"
};

const styleCardContent = {
    maxHeight: "48vh",
    overflow: "auto"
};

/**
 * Component that shows Repositories that isn't Project.
 * 
 * @author Thaynan Nunes
 */
export class RepositoriesNoProject extends Component {
    constructor(props) {
        super(props);

        this.state = {
            ownerName: "",
            repositoryName: ""
        };

        this.addRepository = this.addRepository.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.searchExternalRepository = this.searchExternalRepository.bind(this);
        this.resetExternalRepositories = this.resetExternalRepositories.bind(this);
    }

    /**
     * Function loaded just after the rendering of the component, in which
     * all Repositories linked to the Project will be obtained.
     */
    componentDidMount() {
        this.props.getRepositoriesNoProject();
    }

    /**
     * Adds the repository into Project.
     * 
     * @param {Object} repository
     *      Repository to be added
     */
    async addRepository(repository) {
        await this.props.addRepositoryInProject(repository);

        toast.success("Repositório adicionado!", {
            position: "top-right",
            autoClose: 2500,
            hideProgressBar: true,
            closeOnClick: true,
            pauseOnHover: true,
            draggable: true
          }); 
    }

    /**
     * Changes the component's state from event.
     */
    handleChange = (event) => {
        this.setState({
            [event.target.name]: event.target.value
        });
    }

    /**
     * Searches by repositories that isn't shared with user.
     */
    searchExternalRepository() {
        this.props.searchExternalRepository(this.state.ownerName, this.state.repositoryName);
    }

    /**
     * Resets external repositories' array.
     */
    resetExternalRepositories() {
        this.props.resetExternalRepositories();
    }

    render() {
        const repositoryElements = (repo) => repo.map((repository, indice) => (
            <tr key={indice}>
                <td>{ repository.name }</td>
                <td>{ repository.creationDate }</td>
                <td>{ repository.owner.name }</td>
                <td>
                    <Tooltip content="Adicionar" style={{whiteSpace: "nowrap"}}>
                        <i className="material-icons left green-text" style={{cursor: "pointer"}}
                            onClick={() => this.addRepository(repository)}>
                            add_box
                        </i>
                    </Tooltip>
                </td>
            </tr>
        ));
        
        const repositoryExternalElements = this.props.msgExternalRepositories !== ""
            ? <div className="card-action" style={styleCardContent}>{this.props.msgExternalRepositories}</div>
            : (<div className="card-action" style={styleCardContent}>
                <table className="striped highlight responsive-table">
                    <thead>
                        <tr>
                            <th>Repositório</th>
                            <th>Data de Criação</th>
                            <th>Responsável</th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>
                        {repositoryElements(this.props.externalRepositories)}
                    </tbody>
                </table>
            </div>);

        return (
            <div style={styles}>
                
                <h3>Adicionar Repositórios ao Projeto</h3>
            
                <ToastContainer />
                
                <div style={{display: "flex", marginBottom: "15px", justifyContent: "flex-end"}}>
                    <Link to="/">
                        <div className="col right">
                            <button className="btn waves-effect waves-light green darken-2"
                                    type="button" name="voltar"
                                    onClick={this.resetExternalRepositories}>
                                Voltar
                                <i className="material-icons left">arrow_back</i>
                            </button>
                        </div>
                    </Link>
                </div>

                <div className="card">
                    <div className="card-content" style={{paddingBottom: "0", paddingTop: "20px"}}>
                        <span className="card-title">Buscar repositórios externos</span>
                        <div className="row" style={{marginBottom: "0"}}>
                            <form className="col s12">
                                <div className="row" style={{marginBottom: "0"}}>
                                    <div className="input-field col s5">
                                        <i className="material-icons prefix">account_circle</i>
                                        <input name="ownerName" type="text" placeholder="Responsável" 
                                               onChange={this.handleChange}/>
                                    </div>
                                    <div className="input-field col s5">
                                        <i className="material-icons prefix">insert_drive_file</i>
                                        <input name="repositoryName" type="tel" 
                                               placeholder="Repositório (Opicional)"
                                               onChange={this.handleChange} />
                                    </div>
                                    <div className="input-field col s2">
                                        <button className="btn waves-effect waves-light blue darken-2"
                                                type="button" name="cadastrar-pr"
                                                onClick={this.searchExternalRepository}>
                                            Buscar
                                        </button>
                                    </div>
                                </div>
                            </form>
                        </div>
                    </div>
                    {this.props.externalRepositories && repositoryExternalElements}
                </div>
                
                <div className="card">
                    <div className="card-content" style={{paddingBottom: "0", paddingTop: "20px"}}>
                        <span className="card-title">Repositórios compartilhados</span>
                    </div>
                    <div className="card-action" style={styleCardContent}>
                        <table className="striped highlight responsive-table">
                            <thead>
                                <tr>
                                    <th>Repositório</th>
                                    <th>Data de Criação</th>
                                    <th>Responsável</th>
                                    <th></th>
                                </tr>
                            </thead>
                            <tbody>
                                {repositoryElements(this.props.repositoriesNoProject)}
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        )
    };
}

const mapStateToProps = (state) => ({
    repositoriesNoProject: state.repositories.repositoriesNoProject,
    externalRepositories: state.repositories.externalRepositories,
    msgExternalRepositories: state.repositories.msgExternalRepositories
});

const mapDispatchToProps = (dispatch) => ({
    getRepositoriesNoProject: () => dispatch(getRepositoriesNoProject()),
    addRepositoryInProject: repository => dispatch(addRepositoryInProject(repository)),
    searchExternalRepository: (ownerName, repositoryName) => dispatch(searchExternalRepository(ownerName, repositoryName)),
    resetExternalRepositories: () => dispatch(resetExternalRepositories())
});

export default requireAuthentication(connect(mapStateToProps, mapDispatchToProps)(RepositoriesNoProject));