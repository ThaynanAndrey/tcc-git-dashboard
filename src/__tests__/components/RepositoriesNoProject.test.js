import React from 'react';
import { shallow, configure } from "enzyme";
import toJson from 'enzyme-to-json';

import * as auth from '../../store/actions/authAction';

import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

import RepositoriesNoProject from '../../components/repository/RepositoriesNoProject';

describe('Test RepositoriesNoProject', () => {
  beforeEach(() => {
    const spy = jest.spyOn(auth, 'isAuthenticated');
    spy.mockReturnValue(true);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly a component to show Repositories are not in Project', () => {
    const [getRepositoriesNoProject, addRepositoryInProject,
            searchExternalRepository, resetRepositoriesNoProject] = new Array(4).fill(jest.fn());
    const idProject = '123';
    const repositoriesNoProject = [];
    const externalRepositories = [];
    const msgExternalRepositories = '';

    const tree = shallow(
      <RepositoriesNoProject 
        getRepositoriesNoProject={getRepositoriesNoProject} addRepositoryInProject={addRepositoryInProject}
        searchExternalRepository={searchExternalRepository} resetRepositoriesNoProject={resetRepositoriesNoProject}
        idProject={idProject} repositoriesNoProject={repositoriesNoProject}
        externalRepositories={externalRepositories} msgExternalRepositories={msgExternalRepositories}
      />
    );

    expect(toJson(tree)).toMatchSnapshot();
  });
})