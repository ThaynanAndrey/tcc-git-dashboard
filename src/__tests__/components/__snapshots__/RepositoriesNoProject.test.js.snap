// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RepositoriesNoProject renders correctly a component to show Repositories are not in Project 1`] = `
<div>
  <Connect(RepositoriesNoProject)
    addRepositoryInProject={[MockFunction]}
    externalRepositories={Array []}
    getRepositoriesNoProject={[MockFunction]}
    idProject="123"
    msgExternalRepositories=""
    repositoriesNoProject={Array []}
    resetRepositoriesNoProject={[MockFunction]}
    searchExternalRepository={[MockFunction]}
  />
</div>
`;
